Unit Field Attribute updates/fixes

maraudical opened the pull request:
Full Changelist:
- Fix to missing object reference error in settings when trying to use the UnitField in a fresh project
- Correctly colors the field when using the [Color] attribute
- Updates visual value when the property value is changed by script
- Now works with any type (even custom) that implements Visual Elements and has numerical fields
- Adjusted unit display to align with text
- Added percent units to the Unit Converter
- Added non-conversion UnitField constructor (for if you just want unit display and no conversion)

maraudical commented:
@v0lt13 Added a couple small changes. Just a comparison of what I was talking about with the Color Attribute being different when also using the Unit Attribute. I guess you could write a change to specifically account for that but by doing a copy of all style variables it future proofs it against other attributes that change anything about styles as well.
![Unit Coloring Issue](https://github.com/user-attachments/assets/32bf603c-199a-4a62-81a0-4e35ecca21c7)

maraudical commented:
Ok pushed changes for most things you mentioned. Just a couple notes:

1. `CopyStyle` is useful when combining multiple attributes, it also reduces the need to anticipate specific functionality like with the color attribute. Even if you had specific checks for other attributes to make sure they work in tandem, what if a user created attribute offsets a field's position or changes a color etc. Additionally if they use some attributes on a custom property drawer with unique animations or other tweaks, those will get lost when creating the duplicate field. I actually do think this would be useful for other attributes in the package, but I want to double check with you one last time before deleting that method.
2. Custom numerical types is a bit misleading, I just meant that it will now work with any custom property drawers. For example you can combo the `[MinMaxSlider]` attribute with the `[UnitField]` attribute and it will work perfectly. If a user has any custom property drawers, they can still add the attribute to their type and it will just search the visual element hierarchy for numeric text fields to replace.
<img width="521" height="26" alt="image" src="https://github.com/user-attachments/assets/9defaacd-7d45-467b-8723-9eec167c7bdd" />

maraudical commented:
Also, it seems like Temperature units are missing from the options in `UnitField` was this intentional?

v0lt13 commented:
Ok, in case of copy style, that sounds useful, so in that case move the function to property drawer base and add it to the documentation then use of it in every attribute drawer that could make use of it.

For the custom property drawers, I see now, ok in that case keep it.

What is the purpose of the CreateFieldCopy function? Before I saw you were using the tracking thing, but now why not use the CreateField function?

Temperature conversion uses offsets, doesn't really work with the current system which uses multiplying.

maraudical commented:
I can see why temperature is a bit harder then.

The CreateFieldCopy is mainly just a wrapper for the CreateFieldForType base method. Because of the way I need to get what type of NumericField it is I can't just outright use CreateFieldForType. Because I am really checking `if (element is FloatField) then CreateFieldForType<float>()` for each numeric type.

I guess I could instead just do
```cs
Type fieldType = null;

if (element is FloatField)
    fieldType = typeof(float);
else if (element is DoubleField)
    fieldType = typeof(double);
else if (element is IntegerField)
    fieldType = typeof(int);
else if (element is UnsignedIntegerField)
    fieldType = typeof(uint);
else if (element is LongField)
    fieldType = typeof(long);
else if (element is UnsignedLongField)
    fieldType = typeof(ulong);

var elementCopy = CreateFieldForType(fieldType, /*Not quite sure how I should get the label here*/, ((double)GetFieldValue(element)) / convertedUnit.conversion, property.hasMultipleDifferentValues);
RegisterValueChangedCallbackByType(fieldType, elementCopy, (changeEvent) => SetFieldValue(element, (double)GetFieldValue(elementCopy) * GetConversion().conversion, true));
element.TrackPropertyValue(property, (property) => SetFieldValue(elementCopy, (double)GetFieldValue(element) / GetConversion().conversion));
```

v0lt13 commented:
In that case feel free to extend the actual create field function to support specific numeric types, but also make sure you update the other functions as well

maraudical commented:
Alright I moved `CopyStyle` to `PropertyDrawerBase`. I didn't want to change anything in the `CreateFieldForType` methods so I just added a few helper methods like `GetFieldType` and `GetFieldLabel`.

`GetFieldType` just checks if the VisualElement is one of the specified UI Element Fields and then returns the base type (for example if the field is an `IntegerField` it will return `int`). I just added a flag to only return it if they fit a specific group, for example numeric types or string types etc.

GetFieldLabel is a carbon copy of GetFieldValue but it returns the string label instead.

Overall the UnitFieldDrawer went from 200+ lines to about 100.
